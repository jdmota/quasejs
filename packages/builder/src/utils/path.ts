const path = require( "path" );
const slash = require( "slash" );

export const reExt = /\.(.+)$/;

export function lowerPath( p: string ) {
  return path.sep === "\\" ? p.toLowerCase() : p;
}

export function relative( id: string, cwd: string ): string {
  return slash( lowerPath( path.relative( cwd, id ) ) );
}

export function relativeURL( id: string, cwd: string ): string {
  return slash( lowerPath( path.relative( path.dirname( cwd ), id ) ) );
}

export function resolvePath( id: string, cwd: string ): string {
  return lowerPath( path.resolve( cwd, id ) );
}

export function getType( id: string ): string {
  const match = id.match( reExt );
  return match ? match[ 1 ] : "";
}

export function makeAbsolute( file: string ) {
  return lowerPath( path.resolve( process.cwd(), file ) );
}

export function isAbsolute( file: string ): boolean {
  return path.isAbsolute( file );
}
