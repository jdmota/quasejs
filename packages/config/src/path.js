// @flow
import { formatOption } from "./formating";

export default class Path {

  +arr: string[];
  where: ?string;

  constructor() {
    this.arr = [];
    this.where = null;
  }

  push( key: string | number ) {
    this.arr.push( key + "" );
  }

  pop() {
    this.arr.pop();
  }

  size() {
    return this.arr.length;
  }

  chainToString() {
    return this.arr.join( "." );
  }

  format() {
    return formatOption( `${this.arr.join( "." )}${this.where ? ` [in ${this.where}]` : ""}` );
  }

}
