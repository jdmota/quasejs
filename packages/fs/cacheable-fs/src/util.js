// @flow

const pathUrl = require( "@quase/path-url" );
const path = require( "path" );

export function dirname( p: string ): string {
  return pathUrl.lowerPath( path.dirname( p ) );
}

export function makeAbsolutePath( p: string ): string {
  return pathUrl.lowerPath( pathUrl.makeAbsolutePath( p ) );
}
