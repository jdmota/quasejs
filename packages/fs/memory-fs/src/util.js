const pathUrl = require( "@quase/path-url" );
const path = require( "path" );

function lower( p ) {
  return path.sep === "\\" ? p.toLowerCase() : p;
}

export function dirname( p ) {
  return lower( path.dirname( p ) );
}

export function makeAbsolute( p ) {
  return pathUrl.isUrl( p ) ? pathUrl.makeAbsoluteUrl( p ) : makeAbsolutePath( p );
}

export function makeAbsolutePath( p ) {
  return lower( pathUrl.makeAbsolutePath( p ) );
}

export function makeAbsoluteUrl( p ) {
  return pathUrl.makeAbsoluteUrl( p );
}

export const isUrl = pathUrl.isUrl;
