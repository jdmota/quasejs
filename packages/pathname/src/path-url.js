import req from "../../_helper/require";

const path = req( "path" );
const URL = typeof window !== "undefined" && window.URL; // eslint-disable-line no-undef
const LOCATION = typeof window !== "undefined" && window.location; // eslint-disable-line no-undef

export function makeAbsolute( file ) {
  return URL ? new URL( file, LOCATION ).href : path.resolve( file );
}

export function resolveAsUrl( from, to ) {
  return URL ? new URL( to, from ).href : path.resolve( path.dirname( from ), to );
}

export function normalize( file ) {
  if ( URL ) {
    return new URL( file, LOCATION ).pathname;
  }
  return path.relative( process.cwd(), file ).replace( /\\/g, "/" );
}
