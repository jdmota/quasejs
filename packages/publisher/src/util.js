require( "any-observable/register/rxjs-all" ); // eslint-disable-line import/no-unassigned-import
const { merge } = require( "rxjs" );
const { filter } = require( "rxjs/operators" );
const streamToObservable = require( "@samverschueren/stream-to-observable" );
const split = require( "split" );
const execa = require( "execa" );
const Listr = require( "listr" );
const issueRegex = require( "issue-regex" );
const terminalLink = require( "terminal-link" );

export function exec( cmd, args, opts ) {
  const cp = execa( cmd, args, opts );

  return merge(
    streamToObservable( cp.stdout.pipe( split() ), { await: cp } ),
    streamToObservable( cp.stderr.pipe( split() ), { await: cp } )
  ).pipe( filter( Boolean ) );
}

export function l( tasks ) {
  return new Listr( tasks || [], {
    // showSubtasks: false
  } );
}

export function error( message ) {
  const err = new Error( message );
  err.__generated = true;
  return err;
}

export function linkifyCompare( url, a, b ) {
  if ( !url ) {
    return `${a}...${b}`;
  }
  if ( !terminalLink.isSupported ) {
    return `${url}/compare/${a}...${b}`;
  }
  return terminalLink( `${a}...${b}`, `${url}/compare/${a}...${b}` );
}

// Adapted from https://github.com/sindresorhus/np

export function linkifyIssues( url, message ) {
  if ( !url || !terminalLink.isSupported ) {
    return message;
  }
  return message.replace( issueRegex(), issue => {
    const issuePart = issue.replace( "#", "/issues/" );
    if ( issue.startsWith( "#" ) ) {
      return terminalLink( issue, `${url}${issuePart}` );
    }
    const host = url.replace( /\.com\/[^]*$/, ".com/" );
    return terminalLink( issue, `${host}${issuePart}` );
  } );
}

export function linkifyCommit( url, commit ) {
  if ( !url || !terminalLink.isSupported ) {
    return commit;
  }
  return terminalLink( commit, `${url}/commit/${commit}` );
}
