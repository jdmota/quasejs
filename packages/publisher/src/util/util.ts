import execa from "execa";
import logSymbols from "log-symbols";
import terminalLink from "terminal-link";
import issueRegex from "issue-regex";
import History from "../history";

const { filter } = require( "rxjs/operators" );
const streamToObservable = require( "@samverschueren/stream-to-observable" );
const split = require( "split" );
const Listr = require( "listr" );

type ExecOpts = {
  cwd?: string;
  history?: History;
  preferLocal?: boolean;
};

export function errorIgnoreStack( error: any ) {
  error.__generated = true;
  throw error;
}

export async function execPromise( cmd: string, args: string[], opts: ExecOpts = {} ) {
  const { history } = opts;
  const operation = [ cmd ].concat( args );
  try {
    if ( history ) history.start( operation );
    const cp = await execa( cmd, args, {
      preferLocal: true,
      ...opts
    } );
    if ( history ) history.end( operation );
    return cp;
  } catch ( error ) {
    error.__generated = true;
    throw error;
  }
}

export async function execStdout( cmd: string, args: string[], opts: ExecOpts = {} ): Promise<string> {
  const { history } = opts;
  const operation = [ cmd ].concat( args );
  try {
    if ( history ) history.start( operation );
    const { stdout } = await execa( cmd, args, {
      preferLocal: true,
      ...opts
    } );
    if ( history ) history.end( operation );
    return stdout;
  } catch ( error ) {
    error.__generated = true;
    throw error;
  }
}

export function execObservable( cmd: string, args: string[], opts: ExecOpts = {} ) {
  const { history } = opts;
  const operation = [ cmd ].concat( args );
  const cp = execa( cmd, args, {
    preferLocal: true,
    ...opts
  } );

  if ( history ) {
    history.start( operation );
    cp.then( () => history.end( operation ) );
  }

  const { all } = cp as any;
  return streamToObservable( all.pipe( split() ) ).pipe( filter( Boolean ) );
}

export function l( tasks?: any[] ) {
  return new Listr( tasks || [], {
    // showSubtasks: false
  } );
}

export function error( message: string ) {
  const err = new Error( message );
  // @ts-ignore
  err.__generated = true;
  return err;
}

export function linkifyCompare( url: string, a: string, b: string ): string {
  if ( !url ) {
    return `${a}...${b}`;
  }
  if ( !terminalLink.isSupported ) {
    return `${url}/compare/${a}...${b}`;
  }
  return terminalLink( `${a}...${b}`, `${url}/compare/${a}...${b}` );
}

// Adapted from https://github.com/sindresorhus/np

export function linkifyIssues( url: string, message: string ): string {
  if ( !url || !terminalLink.isSupported ) {
    return message;
  }
  return message.replace( issueRegex(), issue => {
    const issuePart = issue.replace( "#", "/issues/" );
    if ( issue.startsWith( "#" ) ) {
      return terminalLink( issue, `${url}${issuePart}` );
    }
    const host = url.replace( /\.com\/[^]*$/, ".com/" );
    return terminalLink( issue, `${host}${issuePart}` );
  } );
}

export function linkifyCommit( url: string, commit: string ): string {
  if ( !url || !terminalLink.isSupported ) {
    return commit;
  }
  return terminalLink( commit, `${url}/commit/${commit}` );
}

export function linkifyCommitRange( url: string, commitRange: string ): string {
  if ( !url || !terminalLink.isSupported ) {
    return commitRange;
  }
  return terminalLink( commitRange, `${url}/commit/${commitRange}` );
}

export function info( message: string ) {
  // eslint-disable-next-line no-console
  console.log( `${logSymbols.info} ${message}` );
}
