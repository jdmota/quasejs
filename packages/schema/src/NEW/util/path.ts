export class Path {
  public readonly parent: Path | null;
  public readonly key: string | number | null;
  public readonly context: string | null;

  constructor(
    parent: Path,
    key: string | number | null,
    context: string | null
  );
  constructor(parent: null, key: null, context: null);
  constructor(
    parent: Path | null,
    key: string | number | null,
    context: string | null
  ) {
    this.parent = parent;
    this.key = key;
    this.context = context;
  }

  isRoot() {
    return this.parent == null;
  }

  push(key: string | number | null, context: string | null = null) {
    return new Path(this, key, context);
  }

  pop(): Path {
    if (this.parent == null)
      throw new Error("Invariant: validation path - cannot pop the root");
    return this.parent;
  }

  static create() {
    return new Path(null, null, null);
  }
}
