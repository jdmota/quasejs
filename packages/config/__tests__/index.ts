import { expect, it } from "@jest/globals";
import path from "path";
import { prettify } from "../../util/path-url";
import { getConfig } from "../index";

const fixturesFolder = path.resolve(import.meta.dirname, "__fixtures__");

it("get config non-multiple level", async () => {
  const results = await getConfig({
    sources: [
      {
        files: "quase-config",
        extensions: ["ts", "mts", "cts", "js", "mjs", "cjs", "json", ""],
      },
      {
        files: "package.json",
        extensions: [],
        rewrite({ config }) {
          return (config as any)["my-key"];
        },
      },
      {
        files: "non-existent-file",
      },
    ],
    findUpOpts: {
      cwd: fixturesFolder,
      stopAt: process.cwd(),
    },
  });

  expect(
    results.map(r => {
      return {
        ...r,
        filename: prettify(r.filename),
      };
    })
  ).toMatchSnapshot();
});

it("get config multiple level", async () => {
  const results = await getConfig({
    sources: [
      {
        files: "quase-config",
        extensions: ["ts", "mts", "cts", "js", "mjs", "cjs", "json", ""],
      },
      {
        files: "package.json",
        extensions: [],
        rewrite({ config }) {
          return (config as any)["my-key"];
        },
      },
      {
        files: "non-existent-file",
      },
    ],
    findUpOpts: {
      cwd: fixturesFolder,
      stopAt: process.cwd(),
      multipleAtLevel: true,
    },
  });

  expect(
    results.map(r => {
      return {
        ...r,
        filename: prettify(r.filename),
      };
    })
  ).toMatchSnapshot();
});
