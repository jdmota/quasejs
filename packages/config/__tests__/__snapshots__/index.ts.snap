// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`get config multiple level 1`] = `
[
  {
    "config": {
      "iAmTheConfigFile": "yes",
    },
    "filename": "packages/config/__tests__/__fixtures__/quase-config.js",
    "ok": true,
    "source": {
      "extensions": [
        "ts",
        "mts",
        "cts",
        "js",
        "mjs",
        "cjs",
        "json",
        "",
      ],
      "files": "quase-config",
    },
  },
  {
    "config": {
      "dependency": 10,
      "iAmTheConfigFile2": "yes",
    },
    "filename": "packages/config/__tests__/__fixtures__/quase-config.mjs",
    "ok": true,
    "source": {
      "extensions": [
        "ts",
        "mts",
        "cts",
        "js",
        "mjs",
        "cjs",
        "json",
        "",
      ],
      "files": "quase-config",
    },
  },
  {
    "config": [Function],
    "filename": "packages/config/__tests__/__fixtures__/quase-config.cjs",
    "ok": true,
    "source": {
      "extensions": [
        "ts",
        "mts",
        "cts",
        "js",
        "mjs",
        "cjs",
        "json",
        "",
      ],
      "files": "quase-config",
    },
  },
  {
    "config": {
      "configFromPkg": "yes",
    },
    "filename": "packages/config/__tests__/__fixtures__/package.json",
    "ok": true,
    "source": {
      "extensions": [],
      "files": "package.json",
      "rewrite": [Function],
    },
  },
]
`;

exports[`get config non-multiple level 1`] = `
[
  {
    "config": {
      "iAmTheConfigFile": "yes",
    },
    "filename": "packages/config/__tests__/__fixtures__/quase-config.js",
    "ok": true,
    "source": {
      "extensions": [
        "ts",
        "mts",
        "cts",
        "js",
        "mjs",
        "cjs",
        "json",
        "",
      ],
      "files": "quase-config",
    },
  },
  {
    "config": {
      "configFromPkg": "yes",
    },
    "filename": "packages/config/__tests__/__fixtures__/package.json",
    "ok": true,
    "source": {
      "extensions": [],
      "files": "package.json",
      "rewrite": [Function],
    },
  },
]
`;
